src/Evolution/Systems/Cce/LinearSolve.cpp

```diff
@@ -0,0 +1,139 @@
+//  sum_n (A * m)_n P_n = sum_n (B * m)_n d_x P_n
+// So, in particular, we can take advantage of this to obtain a formula for the
+// coefficients of the integral given coefficients m_1 of the input:
+//   sum_n (m_1)_n P_n = sum_n (B * A^{-1} * m_1)_n d_x P_n,
+// Therefore the matrix we wish to act with is B * A^{-1}.
+// In this function we calculate the matrix M such that
+//   sum_n (m_2)_n P_n = sum_n (M * m_2) ((1 - x) d_x P_n  + 2 * P_n)
+// Using
+//   sum_n  m_n  (1 + x) P_n  = sum_n (C * m)_n P_n,
+// so
+//  sum_n (C * A * m)_n P_n = sum_n (B * m)_n ((1 - x) d_x P_n)
+// Adding sum_n (2* B * m_n P_n) to both sides,
+//   sum_n ((C * A + 2 B ) * m)_n P_n
+//        = sum_n (B * m)_n ((1 - x) d_x P_n + 2 * P_n)
+// Therefore, the matrix M = B * (C * A + 2 B)^-1 .
+Matrix q_integration_matrix(const size_t number_of_points) noexcept {
+  Matrix inverse_one_minus_y = Matrix(number_of_points, number_of_points, 0.0);
+  for (size_t i = 1; i < number_of_points - 1; ++i) {
+    double n = i;
+    inverse_one_minus_y(i, i - 1) = -n / (2.0 * n - 1.0);
+    inverse_one_minus_y(i, i) = 1.0;
+    inverse_one_minus_y(i, i + 1) = -(n + 1.0) / (2.0 * n + 3.0);
+  }
+  inverse_one_minus_y(0, 0) = 1.0;
+  inverse_one_minus_y(0, 1) = -1.0 / 3.0;
+  inverse_one_minus_y(number_of_points - 1, number_of_points - 2) =
+      -(number_of_points - 1.0) / (2.0 * (number_of_points - 1.0) - 1.0);
+  inverse_one_minus_y(number_of_points - 1, number_of_points - 1) = 1.0;
+
+  Matrix indef_int(number_of_points, number_of_points, 0.0);
+  for (size_t i = 1; i < number_of_points - 1; ++i) {
+    indef_int(i, i - 1) = 1.0;
+    indef_int(i, i + 1) = -1.0;
+  }
+  indef_int(0, 1) = -1.0;
+  indef_int(number_of_points - 1, number_of_points - 2) = 1.0;
+
+  Matrix dy_identity_lhs(number_of_points, number_of_points, 0.0);
+  for (size_t i = 0; i < number_of_points - 1; ++i) {
+    dy_identity_lhs(i, i) = 2.0 * i + 1.0;
+  }
+
+  Matrix lhs_mat = inverse_one_minus_y * dy_identity_lhs;
+
+  for (size_t i = 1; i < number_of_points - 1; ++i) {
+    lhs_mat(i, i - 1) += 2.0;
+    lhs_mat(i, i + 1) += -2.0;
+  }
+  lhs_mat(0, 1) += -2.0;
+  lhs_mat(number_of_points - 1, number_of_points - 2) += 2.0;
+
+  return Spectral::modal_to_nodal_matrix<Spectral::Basis::Legendre,
+                                         Spectral::Quadrature::GaussLobatto>(
+             number_of_points) *
+         indef_int * inv(lhs_mat) *
+         Spectral::nodal_to_modal_matrix<Spectral::Basis::Legendre,
+                                         Spectral::Quadrature::GaussLobatto>(
+             number_of_points);
+}
+}  // namespace
+
+const Matrix& precomputed_cce_q_integrator(
+    const size_t number_of_radial_grid_points) noexcept {
+  const auto lazy_matrix_cache = make_static_cache<CacheRange<
+      1, Spectral::maximum_number_of_points<Spectral::Basis::Legendre> + 1>>(
+      [](const size_t generator_number_of_radial_points) noexcept {
```

Comment:
we prefix variable names that would otherwise shadow another variable name with `local_` elsewhere in the code, so please name this `local_number_of_radial_points `